### ISSUE TITLE
Adding support for CatBoost Classifier

### ISSUE THREAD

**Issue by pwittchen** [jpmml/sklearn2pmml-plugin#11]
Hi,

I want to use CatBoost classifier in my project. I noticed that [sklearn2pmml](https://github.com/jpmml/sklearn2pmml) does not support this classifier at this moment. I know that there is project [jpmml-catboost](https://github.com/jpmml/jpmml-catboost), which converts CBM to PMML, but in my case, I don't have CBM model and I want to perform prediction with PMML model.

I know, CatBoost by default returns Boolean value for prediction in the output, but I need to get continuous value of type `Double` from 0 to 1. I'm trying to write a Classifier, which will support CatBoost PMML model. I know that I need to get `approx` value from the PMML in the output and generate continuous score of type `Double` from 0 to 1 and classification needs to be done with `LOGIT` algorithm.

Here's what I've got so far. I suppose this code is wrong or incomplete.

```java
public class CatBoost extends Classifier {
    public CatBoost(String module, String name) {
        super(module, name);
    }

    @Override
    public Model encodeModel(Schema schema) {
        ContinuousLabel label = new ContinuousLabel(null, DataType.DOUBLE);
        List<? extends Feature> features = schema.getFeatures();
        Schema catBoostSchema = new Schema(new PMMLEncoder(), label, features);
        
        // I think line below is wrong, but I don't know how to create or get this Model object properly
        Model catBoostModel = catBoostSchema.getEncoder().encodePMML().getModels().get(0);

        OutlierTransformation outlier = new OutlierTransformation() {
            @Override
            public Expression createExpression(FieldRef fieldRef) {
                return PMMLUtil.createApply(PMMLFunctions.LESSOREQUAL, fieldRef, PMMLUtil.createConstant(0d));
            }
        };

        FieldName approxField = FieldName.create("approx");
        Output approx = ModelUtil.createPredictedOutput(
                approxField, OpType.CONTINUOUS, DataType.DOUBLE, outlier
        );

        catBoostModel.setOutput(approx);

        return MiningModelUtil.createBinaryLogisticClassification(
                catBoostModel, 1d, 0d, RegressionModel.NormalizationMethod.LOGIT, true, catBoostSchema
        );
    }
}
```

and added the following line in the `sklearn2pmml.properties` file in the Python project, which uses the plugin:

```
catboost.core.CatBoostClassifier = digital.fingerprints.CatBoost
```

Here's an example of my CatBoost PMML model:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<PMML version="4.3" xmlns="http://www.dmg.org/PMML-4_3" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
<Header>
<Application name="CatBoost" version="Git info:
    Commit: a17d5edd9e55246766fc33a3243ac6a2b0b53acc
    Branch: heads/master
    Author: arcadia-devtools &lt;arcadia-devtools@yandex-team.ru&gt;
    Summary: Solution files and make files updated
    git-svn info:
    Last Changed Rev: 8047363

Other info:
    Build by: go
    Top src dir: /var/lib/go-agent/pipelines/BuildMaster/catboost.git
    Top build dir: /storage/.ya/build
    Hostname: catboost-builder-linux
    Host information: 
        Linux catboost-builder-linux 4.15.0-88-generic #88-Ubuntu SMP Tue Feb 11 20:11:34 UTC 2020 x86_64

    "/>
<Timestamp>2021-05-11T09:08:06Z</Timestamp>
</Header>
<DataDictionary>
<DataField name="values_0" optype="continuous" dataType="float"/>
<DataField name="values_1" optype="continuous" dataType="float"/>
<DataField name="values_2" optype="continuous" dataType="float"/>
<DataField name="values_3" optype="continuous" dataType="float"/>
<DataField name="values_4" optype="continuous" dataType="float"/>
<DataField name="values_5" optype="continuous" dataType="float"/>
<DataField name="values_6" optype="continuous" dataType="float"/>
<DataField name="values_7" optype="continuous" dataType="float"/>
<DataField name="prediction" optype="categorical" dataType="boolean"/>
<DataField name="approx" optype="continuous" dataType="double"/>
</DataDictionary>
<MiningModel functionName="classification">
<MiningSchema>
<MiningField name="values_0" usageType="active"/>
<MiningField name="values_1" usageType="active"/>
<MiningField name="values_2" usageType="active"/>
<MiningField name="values_3" usageType="active"/>
<MiningField name="values_4" usageType="active"/>
<MiningField name="values_5" usageType="active"/>
<MiningField name="values_6" usageType="active"/>
<MiningField name="values_7" usageType="active"/>
<MiningField name="prediction" usageType="target"/>
</MiningSchema>
<Segmentation multipleModelMethod="modelChain">
<Segment id="treeEnsemble">
<True/>
<MiningModel functionName="regression">
<MiningSchema>
<MiningField name="values_0" usageType="active"/>
<MiningField name="values_1" usageType="active"/>
<MiningField name="values_2" usageType="active"/>
<MiningField name="values_3" usageType="active"/>
<MiningField name="values_4" usageType="active"/>
<MiningField name="values_5" usageType="active"/>
<MiningField name="values_6" usageType="active"/>
<MiningField name="values_7" usageType="active"/>
</MiningSchema>
<Output>
<OutputField name="approx" optype="continuous" dataType="double"/>
</Output>
<Segmentation multipleModelMethod="sum">
<Segment id="0">
<True/>
<TreeModel modelName="tree_0" functionName="regression" missingValueStrategy="defaultChild" splitCharacteristic="binarySplit">
<MiningSchema>
<MiningField name="values_0" usageType="active"/>
<MiningField name="values_1" usageType="active"/>
<MiningField name="values_2" usageType="active"/>
<MiningField name="values_3" usageType="active"/>
<MiningField name="values_4" usageType="active"/>
<MiningField name="values_5" usageType="active"/>
<MiningField name="values_6" usageType="active"/>
<MiningField name="values_7" usageType="active"/>
<MiningField name="approx" usageType="target"/>
</MiningSchema>
<Output>
<OutputField name="approx" optype="continuous" dataType="double"/>
</Output>
<Node id="0" recordCount="64403.59863">
<True/>
<Node id="2" recordCount="18597.64966" defaultChild="5">
<SimplePredicate field="values_4" operator="greaterThan" value="0.205"/>
<Node id="6" recordCount="3772.899902" defaultChild="13">
<SimplePredicate field="values_2" operator="greaterThan" value="0.325"/>
<Node id="14" recordCount="866" defaultChild="29">
<SimplePredicate field="values_7" operator="greaterThan" value="0.0171"/>
<Node id="30" score="-0.05783786193" recordCount="651">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0185"/>
</Node>
<Node id="29" score="-0.05389908543" recordCount="215">
<True/>
</Node>
</Node>
<Node id="13" recordCount="2906.899902">
<True/>
<Node id="28" score="-0.05368810541" recordCount="207">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0185"/>
</Node>
<Node id="27" score="0.04334205595" recordCount="2699.899902">
<True/>
</Node>
</Node>
</Node>
<Node id="5" recordCount="14824.74976">
<True/>
<Node id="12" recordCount="4753" defaultChild="25">
<SimplePredicate field="values_7" operator="greaterThan" value="0.0171"/>
<Node id="26" score="-0.05954601343" recordCount="3192">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0185"/>
</Node>
<Node id="25" score="-0.05907895167" recordCount="1561">
<True/>
</Node>
</Node>
<Node id="11" recordCount="10071.74976">
<True/>
<Node id="24" score="-0.004387479327" recordCount="2519.949951">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0185"/>
</Node>
<Node id="23" score="0.01412698855" recordCount="7551.799805">
<True/>
</Node>
</Node>
</Node>
</Node>
<Node id="1" recordCount="45805.94897">
<True/>
<Node id="4" recordCount="504" defaultChild="9">
<SimplePredicate field="values_2" operator="greaterThan" value="0.325"/>
<Node id="10" recordCount="241" defaultChild="21">
<SimplePredicate field="values_7" operator="greaterThan" value="0.0171"/>
<Node id="22" score="-0.05203556241" recordCount="159">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0185"/>
</Node>
<Node id="21" score="-0.04626836491" recordCount="82">
<True/>
</Node>
</Node>
<Node id="9" recordCount="263">
<True/>
<Node id="20" score="-0.04722915524" recordCount="90">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0185"/>
</Node>
<Node id="19" score="-0.0526005988" recordCount="173">
<True/>
</Node>
</Node>
</Node>
<Node id="3" recordCount="45301.94897">
<True/>
<Node id="8" recordCount="12431.94995" defaultChild="17">
<SimplePredicate field="values_7" operator="greaterThan" value="0.0171"/>
<Node id="18" score="-0.0598149642" recordCount="7867">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0185"/>
</Node>
<Node id="17" score="-0.02916858895" recordCount="4564.949951">
<True/>
</Node>
</Node>
<Node id="7" recordCount="32869.99902">
<True/>
<Node id="16" score="-0.0594948086" recordCount="2866">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0185"/>
</Node>
<Node id="15" score="0.03331652964" recordCount="30003.99902">
<True/>
</Node>
</Node>
</Node>
</Node>
</Node>
</TreeModel>
</Segment>
<Segment id="1">
<True/>
<TreeModel modelName="tree_1" functionName="regression" missingValueStrategy="defaultChild" splitCharacteristic="binarySplit">
<MiningSchema>
<MiningField name="values_0" usageType="active"/>
<MiningField name="values_1" usageType="active"/>
<MiningField name="values_2" usageType="active"/>
<MiningField name="values_3" usageType="active"/>
<MiningField name="values_4" usageType="active"/>
<MiningField name="values_5" usageType="active"/>
<MiningField name="values_6" usageType="active"/>
<MiningField name="values_7" usageType="active"/>
<MiningField name="approx" usageType="target"/>
</MiningSchema>
<Output>
<OutputField name="approx" optype="continuous" dataType="double"/>
</Output>
<Node id="0" recordCount="64403.59863">
<True/>
<Node id="2" recordCount="35984.29932" defaultChild="5">
<SimplePredicate field="values_0" operator="greaterThan" value="0.151"/>
<Node id="6" recordCount="29993.44946" defaultChild="13">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0095"/>
<Node id="14" recordCount="29983.44946" defaultChild="29">
<SimplePredicate field="values_4" operator="greaterThan" value="0.011"/>
<Node id="30" score="-0.05830787529" recordCount="9102">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0197667"/>
</Node>
<Node id="29" score="0.01348776275" recordCount="20881.44946">
<True/>
</Node>
</Node>
<Node id="13" recordCount="10">
<True/>
<Node id="28" score="-0.01152217616" recordCount="6">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0197667"/>
</Node>
<Node id="27" score="-0.008454164302" recordCount="4">
<True/>
</Node>
</Node>
</Node>
<Node id="5" recordCount="5990.849854">
<True/>
<Node id="12" recordCount="5980.849854" defaultChild="25">
<SimplePredicate field="values_4" operator="greaterThan" value="0.011"/>
<Node id="26" score="0" recordCount="0">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0197667"/>
</Node>
<Node id="25" score="0.009519921624" recordCount="5980.849854">
<True/>
</Node>
</Node>
<Node id="11" recordCount="10">
<True/>
<Node id="24" score="0" recordCount="0">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0197667"/>
</Node>
<Node id="23" score="-0.01749365779" recordCount="10">
<True/>
</Node>
</Node>
</Node>
</Node>
<Node id="1" recordCount="28419.29932">
<True/>
<Node id="4" recordCount="9405" defaultChild="9">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0095"/>
<Node id="10" recordCount="8881" defaultChild="21">
<SimplePredicate field="values_4" operator="greaterThan" value="0.011"/>
<Node id="22" score="-0.05805384008" recordCount="5764">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0197667"/>
</Node>
<Node id="21" score="-0.05938485359" recordCount="3117">
<True/>
</Node>
</Node>
<Node id="9" recordCount="524">
<True/>
<Node id="20" score="-0.05502273322" recordCount="414">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0197667"/>
</Node>
<Node id="19" score="-0.04892984305" recordCount="110">
<True/>
</Node>
</Node>
</Node>
<Node id="3" recordCount="19014.29932">
<True/>
<Node id="8" recordCount="18577.29932" defaultChild="17">
<SimplePredicate field="values_4" operator="greaterThan" value="0.011"/>
<Node id="18" score="0" recordCount="0">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0197667"/>
</Node>
<Node id="17" score="0.04452435508" recordCount="18577.29932">
<True/>
</Node>
</Node>
<Node id="7" recordCount="437">
<True/>
<Node id="16" score="0" recordCount="0">
<SimplePredicate field="values_5" operator="greaterThan" value="0.0197667"/>
</Node>
<Node id="15" score="-0.05767462022" recordCount="437">
<True/>
</Node>
</Node>
</Node>
</Node>
</Node>
</TreeModel>
</Segment>
<Segment id="2">
<True/>
<TreeModel modelName="tree_2" functionName="regression" missingValueStrategy="defaultChild" splitCharacteristic="binarySplit">
<MiningSchema>
<MiningField name="values_0" usageType="active"/>
<MiningField name="values_1" usageType="active"/>
<MiningField name="values_2" usageType="active"/>
<MiningField name="values_3" usageType="active"/>
<MiningField name="values_4" usageType="active"/>
<MiningField name="values_5" usageType="active"/>
<MiningField name="values_6" usageType="active"/>
<MiningField name="values_7" usageType="active"/>
<MiningField name="approx" usageType="target"/>
</MiningSchema>
<Output>
<OutputField name="approx" optype="continuous" dataType="double"/>
</Output>

<!-- ... -->
<!-- I CUT THE REST OF THE FILE BECASUE GITHUB DOES NOT ACCEPT TOO LONG ISSUES -->
<!-- ... -->

</PMML>
```

Before my changes, I've got the following error:

```
Standard error:
May 17, 2021 3:04:40 PM org.jpmml.python.PickleUtil registerClassDictConstructor
WARNING: Failed to load Java class sklearn2pmml.ensemble.SelectFirstEstimator
May 17, 2021 3:04:40 PM org.jpmml.python.PickleUtil registerClassDictConstructor
WARNING: Failed to load Java class org.jpmml.sklearn.CClassDict
May 17, 2021 3:04:40 PM org.jpmml.python.PickleUtil registerClassDictConstructor
WARNING: Failed to load Java class org.jpmml.sklearn.CClassDict
May 17, 2021 3:04:40 PM org.jpmml.sklearn.Main run
INFO: Parsing PKL..
May 17, 2021 3:04:40 PM org.jpmml.sklearn.Main run
INFO: Parsed PKL in 21 ms.
May 17, 2021 3:04:40 PM org.jpmml.sklearn.Main run
INFO: Converting PKL to PMML..
May 17, 2021 3:04:40 PM org.jpmml.sklearn.Main run
SEVERE: Failed to convert PKL to PMML
java.lang.IllegalArgumentException: The transformer object (Python class catboost.core.CatBoostClassifier) is not a supported Transformer
	at org.jpmml.python.CastFunction.apply(CastFunction.java:45)
	at sklearn.pipeline.Pipeline$1.apply(Pipeline.java:108)
	at sklearn.pipeline.Pipeline$1.apply(Pipeline.java:95)
	at com.google.common.collect.Lists$TransformingRandomAccessList.get(Lists.java:638)
	at sklearn2pmml.pipeline.PMMLPipeline.getHead(PMMLPipeline.java:585)
	at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:198)
	at org.jpmml.sklearn.Main.run(Main.java:233)
	at org.jpmml.sklearn.Main.main(Main.java:151)
Caused by: java.lang.ClassCastException: Cannot cast net.razorvine.pickle.objects.ClassDict to sklearn.Transformer
	at java.lang.Class.cast(Class.java:3369)
	at org.jpmml.python.CastFunction.apply(CastFunction.java:43)
	... 7 more

Exception in thread "main" java.lang.IllegalArgumentException: The transformer object (Python class catboost.core.CatBoostClassifier) is not a supported Transformer
	at org.jpmml.python.CastFunction.apply(CastFunction.java:45)
	at sklearn.pipeline.Pipeline$1.apply(Pipeline.java:108)
	at sklearn.pipeline.Pipeline$1.apply(Pipeline.java:95)
	at com.google.common.collect.Lists$TransformingRandomAccessList.get(Lists.java:638)
	at sklearn2pmml.pipeline.PMMLPipeline.getHead(PMMLPipeline.java:585)
	at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:198)
	at org.jpmml.sklearn.Main.run(Main.java:233)
	at org.jpmml.sklearn.Main.main(Main.java:151)
Caused by: java.lang.ClassCastException: Cannot cast net.razorvine.pickle.objects.ClassDict to sklearn.Transformer
	at java.lang.Class.cast(Class.java:3369)
	at org.jpmml.python.CastFunction.apply(CastFunction.java:43)
	... 7 more
```

After my changes, I've got the following error:

```
May 17, 2021 2:37:53 PM org.jpmml.python.PickleUtil registerClassDictConstructor
WARNING: Failed to load Java class sklearn2pmml.ensemble.SelectFirstEstimator
May 17, 2021 2:37:53 PM org.jpmml.python.PickleUtil registerClassDictConstructor
WARNING: Failed to load Java class org.jpmml.sklearn.CClassDict
May 17, 2021 2:37:53 PM org.jpmml.python.PickleUtil registerClassDictConstructor
WARNING: Failed to load Java class org.jpmml.sklearn.CClassDict
May 17, 2021 2:37:53 PM org.jpmml.sklearn.Main run
INFO: Parsing PKL..
May 17, 2021 2:37:53 PM org.jpmml.sklearn.Main run
INFO: Parsed PKL in 26 ms.
May 17, 2021 2:37:53 PM org.jpmml.sklearn.Main run
INFO: Converting PKL to PMML..
May 17, 2021 2:37:53 PM org.jpmml.sklearn.Main run
SEVERE: Failed to convert PKL to PMML
java.lang.IllegalArgumentException: The value of 'catboost.core.CatBoostClassifier.classes_' attribute (null) is not a supported array type
	at org.jpmml.python.PythonObject.getArray(PythonObject.java:171)
	at sklearn.Classifier.getClasses(Classifier.java:43)
	at sklearn.ClassifierUtil.getClasses(ClassifierUtil.java:32)
	at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:137)
	at org.jpmml.sklearn.Main.run(Main.java:233)
	at org.jpmml.sklearn.Main.main(Main.java:151)

Exception in thread "main" java.lang.IllegalArgumentException: The value of 'catboost.core.CatBoostClassifier.classes_' attribute (null) is not a supported array type
	at org.jpmml.python.PythonObject.getArray(PythonObject.java:171)
	at sklearn.Classifier.getClasses(Classifier.java:43)
	at sklearn.ClassifierUtil.getClasses(ClassifierUtil.java:32)
	at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:137)
	at org.jpmml.sklearn.Main.run(Main.java:233)
	at org.jpmml.sklearn.Main.main(Main.java:151)

Preserved joblib dump file(s): /tmp/pipeline-tqh95pk_.pkl.z
```

which indicates that my custom Classifier was loaded, but it's not able to encode model. I see that the `classes_` attribute is null, but I have no idea, how to update my Classifier to support CatBoost properly. I guess the `classes_` is not the only thing, which is missing in the classifier implementation.

I will appreciate any help.

Regards,
Piotr

**Comment by vruusmann** [comment]
> I want to use CatBoost classifier in my project. I noticed that sklearn2pmml does not support this classifier at this moment.

The main "innovation" of CatBoost is a scheme for encoding categorical features.

This encoding can be emulated by third-party tools, such as the `[category_encoders.cat_boost.CatBoostEncoder](https://contrib.scikit-learn.org/category_encoders/catboost.html)` transformation type.

The `CatBoostEncoder` transformation is well supported by recent versions of SkLearn2PMML.

> I know that there is project jpmml-catboost, which converts CBM to PMML

CatBoost stores models in proprietary binary data format. The JPMML-CatBoost library was my attempt to build a Java parser for this data format. It's a good start, but it didn't make to so far to be able to parse data structures related to categorical features.

> // I think line below is wrong, but I don't know how to create or get this Model object properly
> Model catBoostModel = catBoostSchema.getEncoder().encodePMML().getModels().get(0);

This is where the parsing of CBM data file is supposed to happen. The JPMML-CateBoost library could be very useful there :-)

> I know, CatBoost by default returns Boolean value for prediction in the output, but I need to get continuous value of type Double from 0 to 1.

Do I understand you correctly that you want to take a CatBoost-generated classifier PMML file, and transform it so that it would start outputting `probability(1)` instead of the label?

The PMML document contains the `approx` output field already.

The simplest way would be to define another output field (ex. `logic(approx)`), which applies the logit transform to it using plain old `Apply` elements.

Sure, the same can be accomplished by defining a `RegressionModel@normalizationMethod='logit'` element, but it changes the model structure substantially (need to create a new top-level `MiningModel` element, which hosts the current `MiningModel` element plus the upcoming `RegressionModel` element).

> java.lang.IllegalArgumentException: The transformer object (Python class catboost.core.CatBoostClassifier) is not a supported Transformer

Your `sklearn2pmml.properties` file didn't contain a transformer-to-transformer converter mapping.

> Exception in thread "main" java.lang.IllegalArgumentException: The value of 'catboost.core.CatBoostClassifier.classes_' attribute (null) is not a supported array type

The `CatBoostClassifier.clases_` attribute is not defined (is null).

Has the CatBoost classifier model been fitted? If so, then perhaps it's using a different attribute name than Scikit-Learn classifiers (the `classes_` attribute)?

**Comment by vruusmann** [comment]
TLDR: If you need better encoding for categorical features, then I'd suggest exploring `CateBoostEncoder` (transformer) over the full-blown `CatBoostClassifier` (model).

**Comment by pwittchen** [comment]
Thank you for the reply.

> Do I understand you correctly that you want to take a CatBoost-generated classifier PMML file, and transform it so that it would start outputting probability(1) instead of the label?

Yes. I want to start outputting probability.

> TLDR: If you need better encoding for categorical features, then I'd suggest exploring CateBoostEncoder (transformer) over the full-blown CatBoostClassifier (model).

Can you give me some hints, how can I utilize `CatBoostEncoder` in my case?

When I remove custom classifier and simply add to my `sklearn2pmml.properties` file lines:

```
category_encoders.cat_boost.CatBoostEncoder = category_encoders.CatBoostEncoder
catboost.core.CatBoostClassifier =
```

and then run the tests.

and remove previously added line, then I get the following error:

```
Standard output is empty
Standard error:
May 18, 2021 1:38:36 PM org.jpmml.python.PickleUtil registerClassDictConstructor
WARNING: Failed to load Java class sklearn2pmml.ensemble.SelectFirstEstimator
May 18, 2021 1:38:36 PM org.jpmml.python.PickleUtil registerClassDictConstructor
WARNING: Failed to load Java class org.jpmml.sklearn.CClassDict
May 18, 2021 1:38:36 PM org.jpmml.python.PickleUtil registerClassDictConstructor
WARNING: Failed to load Java class org.jpmml.sklearn.CClassDict
May 18, 2021 1:38:36 PM org.jpmml.sklearn.Main run
INFO: Parsing PKL..
May 18, 2021 1:38:36 PM org.jpmml.sklearn.Main run
INFO: Parsed PKL in 16 ms.
May 18, 2021 1:38:36 PM org.jpmml.sklearn.Main run
INFO: Converting PKL to PMML..
May 18, 2021 1:38:36 PM org.jpmml.sklearn.Main run
SEVERE: Failed to convert PKL to PMML
java.lang.IllegalArgumentException: The transformer object (Python class catboost.core.CatBoostClassifier) is not a supported Transformer
	at org.jpmml.python.CastFunction.apply(CastFunction.java:45)
	at sklearn.pipeline.Pipeline$1.apply(Pipeline.java:108)
	at sklearn.pipeline.Pipeline$1.apply(Pipeline.java:95)
	at com.google.common.collect.Lists$TransformingRandomAccessList.get(Lists.java:638)
	at sklearn2pmml.pipeline.PMMLPipeline.getHead(PMMLPipeline.java:585)
	at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:198)
	at org.jpmml.sklearn.Main.run(Main.java:244)
	at org.jpmml.sklearn.Main.main(Main.java:161)
Caused by: java.lang.ClassCastException: Cannot cast net.razorvine.pickle.objects.ClassDict to sklearn.Transformer
	at java.lang.Class.cast(Class.java:3369)
	at org.jpmml.python.CastFunction.apply(CastFunction.java:43)
	... 7 more

Exception in thread "main" java.lang.IllegalArgumentException: The transformer object (Python class catboost.core.CatBoostClassifier) is not a supported Transformer
	at org.jpmml.python.CastFunction.apply(CastFunction.java:45)
	at sklearn.pipeline.Pipeline$1.apply(Pipeline.java:108)
	at sklearn.pipeline.Pipeline$1.apply(Pipeline.java:95)
	at com.google.common.collect.Lists$TransformingRandomAccessList.get(Lists.java:638)
	at sklearn2pmml.pipeline.PMMLPipeline.getHead(PMMLPipeline.java:585)
	at sklearn2pmml.pipeline.PMMLPipeline.encodePMML(PMMLPipeline.java:198)
	at org.jpmml.sklearn.Main.run(Main.java:244)
	at org.jpmml.sklearn.Main.main(Main.java:161)
Caused by: java.lang.ClassCastException: Cannot cast net.razorvine.pickle.objects.ClassDict to sklearn.Transformer
	at java.lang.Class.cast(Class.java:3369)
	at org.jpmml.python.CastFunction.apply(CastFunction.java:43)
	... 7 more
```

so sklearn2pmml still wants to use `CatBoostClassifier`. I'm not sure, how to use it properly and optionally force it to not use CatBoostClassifier and use CatBoostEncoder with `probability(1)`. I know that I can add custom output fields in the case of Classifier, but I'm not sure if it's possible in case of Encoder/Transformer.

Moreover, when I remove these two lines from the configuration in the properties file, then effect is the same.

**Comment by vruusmann** [comment]
> Can you give me some hints, how can I utilize CatBoostEncoder in my case?

In brief, you should update your Python model training script:

Bad: Using `CatBoostClassifier`
Good: Using `CatBoostEncoder` to handle categorical features, which is then followed by any ordinary decision tree ensemble classifier.

`CatBoostEncoder + GradientBoostingClassifier` is functionally very close to `CatBoostClassifier`.

Works out of the box. No need to implement any custom transformers, or create custom `sklearn2pmml.properties` mappings.

**Comment by pwittchen** [comment]
Ok, thanks. I'll test that proposed solution.
